File: scalaz/src/main/scala/step1_read_print.scala
import types._

import scala.annotation.tailrec
import scala.io.StdIn
import scala.util.control.NonFatal

object step1_read_print {

  def read(str: String): MalType = reader.read_str(str)

  def eval(ast: MalType, env: String): MalType = ast

  def print(ast: MalType): String = ast.show()

  def rep(str: String): String = print(eval(read(str), ""))

  def main(args: Array[String]): Unit = {
    @tailrec
    def go(): Unit = Option(StdIn.readLine("user> ")) match {
      case Some(line) =>
        try println(rep(line)) catch {
          case NonFatal(e) => e.printStackTrace()
        }
        go()
      case None => ()
    }

    go()
  }
}

File: java8/src/main/java/mal/step1_read_print.java
package mal;

import java.io.IOException;
import mal.Types.*;

public class step1_read_print implements Step<MalType, String> {

    public static void main(String[] args) throws IOException {
        step1_read_print instance = new step1_read_print();
        String line;
        while ((line = IO.prompt("user> ")) != null) {
            try {
                System.out.println(instance.rep(line, ""));
            } catch (Exception e) {
                System.err.print("Error: ");
                e.printStackTrace();
            }
        }
    }

    @Override
    public MalType read(String str) {
        return Reader.read_str(str);
    }

    @Override
    public MalType eval(MalType ast, String env) {
        return ast;
    }

    @Override
    public String print(MalType ast) {
        return Printer.pr_str(ast);
    }
}
